### PR TITLE
[perf] Add a Clojure-side impl for `lib.cache/side-channel-cache`

### PR DESCRIPTION
**Pull request by bshepherdson**
This puts a cache into the (Clojure) metadata on each query when it gets created on the JVM side.

Since metadata persists across `assoc` etc., we need to also key on the current value of the query. Therefore the metadata actually contains a `WeakHashMap` using the query map itself as the key. Inside that is an `(atom {})` caching the actual values.

The net result is that the caching for a given query has the same lifespan as references to that query.
That's generally brief but often includes many calls to expensive, cached functions like `visible-columns`.

In my testing this cut the time required for processing `field-usage` on a medium-complexity query (4 filters, 1 aggregation, 1 breakout) from 135ms to 72ms.

**Memory impact**

`WeakHashMap`s have weak references to their keys but strong references to their values. Therefore if some value in the map retains a key, that key will persist. I *think* that isn't an issue here - the keys are queries but the values are things like `visible-columns` results that don't hang onto the whole query.

Still, it would be good to check that this doesn't make the memory usage jump on stats, before it gets released.

### CODE REVIEW

**Comment by bshepherdson** [comment]
* **#45394** <a href="https://app.graphite.dev/github/pr/metabase/metabase/45394?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bshepherdson and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/metabase/lib/cache.cljc [hunk 1]
```diff
@@ -17,12 +43,25 @@
...
 
+(defn attach-query-cache
+  "Attaches the cache to a newly constructed query.
+
+  This uses metadata on CLJ and does nothing on CLJS."
+  [query]
+  #?(:cljs query
+     :clj  (vary-meta query (fn [mmeta]
+                              (cond-> mmeta
+                                (contains? mmeta :lib/__cache)
```
Comment (metamben) [inline src/metabase/lib/cache.cljc:62]: What makes this true?

File: src/metabase/lib/cache.cljc [hunk 2]
```diff
@@ -58,46 +97,49 @@
...
-                   ;; Cache miss - generate the value and cache it.
-                   (let [value (f x)]
-                     (cache-fn cache subkey x value)
-                     value))
-                 (f x)))
-             (f x))))
+  (if (or force?
+          (map? host)
+          #?(:cljs (object? host)
+             :clj  false))
```
Comment (metamben) [inline src/metabase/lib/cache.cljc:103]: Why is the `:clj` branch needed?

**Review by metamben — COMMENTED** [review]
I don't understand how the cache gets attached.

File: src/metabase/lib/cache.cljc (hunk 1 above)
Comment (bshepherdson) [reply src/metabase/lib/cache.cljc]: I actually inverted this logic by accident while refactoring. It's fixed now.

File: src/metabase/lib/cache.cljc (hunk 2 above)
Comment (bshepherdson) [reply src/metabase/lib/cache.cljc:103]: I suppose it's not; `nil` would have worked just as well.

Comment (metamben) [reply src/metabase/lib/cache.cljc]: I expect it to be completely omitted in CLJ if there is no `:clj` clause, so not even a `nil`.

**Review by metamben — APPROVED** [review]
Maybe a test for checking that the cache gets attached and works in CLJ would make sense.

File: src/metabase/lib/cache.cljc (hunk 2 above)
Comment (bshepherdson) [reply src/metabase/lib/cache.cljc]: That's not how `#?` works; it resolves to `nil` in dialects which are not included in the conditional. That's fine here, though.

That's why we use `#?@(:cljs ((:require-macros ...)))` with a one-element list - `nil` isn't allowed in `ns` macros and splicing in `nil` is an empty seq, so it disappears.

Comment (metamben) [reply src/metabase/lib/cache.cljc]: Why do I see
```
user> (count [0 #?(:clj 1) #?(:cljs 2)])
2
```
in my REPLs? Shouldn't it be `3`?

Comment (bshepherdson) [reply src/metabase/lib/cache.cljc]: Huh, I was very confident in that. Maybe there's some context where they work differently, but it certainly does seem like `#?(:cljs 7)` vanishes in CLJ and vice versa. TIL!

**Pull request merged**